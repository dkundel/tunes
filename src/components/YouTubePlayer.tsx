import * as React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import { GlobalState, PlayerStatus, SongInfo } from '../reducers';
import {
  getPlayerStatus,
  getCurrentQueue,
  getCurrentPlayList,
  getQueueIndex,
  getNextPlayerStatus
} from '../reducers/player';
import {
  changePlayerStatus,
  setQueueIndex,
  nextSong,
  setSongInfo,
  setProgress
} from '../actions';

export const YOUTUBE_PLAYER_DOM_ID = 'youTubePlayer';

export interface YouTubePlayerProps {
  currentPlayList?: string | undefined;
  playerStatus?: PlayerStatus;
  currentQueue?: string[] | undefined;
  nextStatus?: PlayerStatus;
  queueIndex?: number | undefined;
  updatePlayerStatus?: (status: PlayerStatus) => any;
  updateQueueIndex?: (index: number) => any;
  nextSong?: () => any;
  setSongInfo?: (info: SongInfo) => any;
  setProgress?: (
    currentTime: number,
    totalTime: number,
    percentage: number
  ) => any;
}

export type YouTubeVideoData = {
  author: string;
  list: string;
  title: string;
  video_id: string;
  video_quality: string;
};

interface extendedYTPlayer extends YT.Player {
  getVideoData(): YouTubeVideoData;
}

type YTOnReadyEvent = { target: extendedYTPlayer };

class YouTubePlayer extends React.Component<YouTubePlayerProps, undefined> {
  private player: extendedYTPlayer | null;
  private timer: NodeJS.Timer | undefined;

  private loadVideo() {
    const defaultOptions: YT.PlayerOptions = {
      width: 300,
      height: 600,
      events: {
        onReady: this.onReady.bind(this),
        onStateChange: this.onStateChange.bind(this),
        onError: this.onError.bind(this)
      },
      videoId: ''
    };

    if (this.props.currentPlayList) {
      if (this.player === null) {
        defaultOptions.playerVars = {
          list: this.props.currentPlayList
        };
        new YT.Player(YOUTUBE_PLAYER_DOM_ID, defaultOptions);
      } else {
        this.player.playVideoAt(this.props.queueIndex || 0);
      }
    } else if (this.props.currentQueue && this.props.queueIndex !== undefined) {
      defaultOptions.videoId = this.props.currentQueue[this.props.queueIndex];
      new YT.Player(YOUTUBE_PLAYER_DOM_ID, defaultOptions);
    }
  }

  private updateTimeOfPlayer() {
    if (!this.player) {
      return;
    }

    if (this.props.setProgress) {
      const currentTime = this.player.getCurrentTime();
      const totalTime = this.player.getDuration();
      const percentage = Math.round(currentTime / totalTime * 10000) / 100.0;
      this.props.setProgress(currentTime, totalTime, percentage);
    }
  }

  private handleTimer(playerState: YT.PlayerState) {
    if (playerState === YT.PlayerState.PLAYING) {
      this.timer = setInterval(() => {
        this.updateTimeOfPlayer();
      }, 1000);
    } else {
      if (this.timer) {
        clearInterval(this.timer);
        this.timer = undefined;
      }
    }
  }

  private onReady(evt: YTOnReadyEvent) {
    if (evt.target) {
      this.player = evt.target;
      this.player.playVideo();
      window['player'] = this.player;
    }
  }

  private onError() {
    // TODO: track blocked videos
    if (this.props.nextSong) {
      this.props.nextSong();
    }
  }

  private onStateChange(evt: YT.OnStateChangeEvent) {
    if (
      !this.player ||
      !this.props.updateQueueIndex ||
      !this.props.updatePlayerStatus ||
      !this.props.nextSong ||
      !this.props.setSongInfo
    ) {
      return;
    }

    this.handleTimer(evt.data);

    /*if (evt.data === YT.PlayerState.UNSTARTED) {
      this.props.updatePlayerStatus('loading');
    } else */
    if (evt.data === YT.PlayerState.PLAYING) {
      this.props.updatePlayerStatus('playing');
      if (this.props.currentPlayList) {
        this.props.updateQueueIndex(this.player.getPlaylistIndex());
      }
      const videoData = this.player.getVideoData();
      const songInfo: SongInfo = {
        currentTime: this.player.getCurrentTime(),
        percentage: 0,
        title: videoData.title,
        videoId: videoData.video_id,
        totalTime: this.player.getDuration()
      };
      this.props.setSongInfo(songInfo);
    } else if (evt.data === YT.PlayerState.PAUSED) {
      this.props.updatePlayerStatus('paused');
    } else if (evt.data === YT.PlayerState.ENDED) {
      this.props.updatePlayerStatus('stopped');
      this.props.nextSong();
    }
  }

  public componentDidMount() {
    this.loadVideo();
  }

  public shouldComponentUpdate(nextProps: YouTubePlayerProps) {
    if (nextProps.nextStatus !== undefined) {
      return true;
    }

    if (
      Array.isArray(nextProps.currentQueue) &&
      Array.isArray(this.props.currentQueue) &&
      nextProps.currentQueue.length !== 0 &&
      nextProps.currentQueue.length === this.props.currentQueue.length &&
      nextProps.playerStatus === this.props.playerStatus &&
      nextProps.queueIndex === this.props.queueIndex
    ) {
      return false;
    }

    return true;
  }

  public componentWillUpdate(nextProps: YouTubePlayerProps) {
    if (nextProps.currentPlayList !== this.props.currentPlayList) {
      this.player = null;
    }
  }

  public componentDidUpdate() {
    if (
      this.props.nextStatus === 'loading' ||
      this.props.nextStatus === 'playing'
    ) {
      this.loadVideo();
    }

    if (this.player) {
      const ytPlayerState = this.player.getPlayerState();
      if (
        (this.props.nextStatus === 'resuming' &&
          ytPlayerState === YT.PlayerState.PAUSED) ||
        ytPlayerState === YT.PlayerState.ENDED
      ) {
        this.player.playVideo();
      } else if (
        this.props.nextStatus === 'paused' &&
        ytPlayerState === YT.PlayerState.PLAYING
      ) {
        this.player.pauseVideo();
      } else if (
        this.props.nextStatus === 'stopped' &&
        ytPlayerState === YT.PlayerState.PLAYING
      ) {
        this.player.stopVideo();
      }
    }
  }

  public render() {
    return <div id={YOUTUBE_PLAYER_DOM_ID} />;
  }
}

const mapStateToProps = (state: GlobalState) => ({
  currentPlayList: getCurrentPlayList(state),
  playerStatus: getPlayerStatus(state),
  currentQueue: getCurrentQueue(state),
  queueIndex: getQueueIndex(state),
  nextStatus: getNextPlayerStatus(state)
});

const mapDispatchToProps = (dispatch: Dispatch<undefined>) => ({
  updatePlayerStatus: (status: PlayerStatus) =>
    dispatch(changePlayerStatus(status)),
  updateQueueIndex: (index: number) => dispatch(setQueueIndex(index)),
  nextSong: () => dispatch(nextSong()),
  setSongInfo: (info: SongInfo) => dispatch(setSongInfo(info)),
  setProgress: (currentTime: number, totalTime: number, percentage: number) =>
    dispatch(setProgress(currentTime, totalTime, percentage))
});

export default connect(mapStateToProps, mapDispatchToProps)(YouTubePlayer);
