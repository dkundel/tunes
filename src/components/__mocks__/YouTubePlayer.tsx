import * as React from 'react';

export default class MockYouTubePlayer extends React.Component<any, any> {
  render() {
    return <div id="YouTubePlayer" />;
  }
}
