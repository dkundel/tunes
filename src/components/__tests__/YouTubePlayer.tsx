type MockPlayerOptions = {
  events: {
    onReady: Function;
    onStateChange: Function;
    onError: Function;
  };
};

class MockPlayer {
  constructor(private domId: string, private opts: MockPlayerOptions) {}
  playVideoAt(): void {}
  getCurrentTime(): void {}
  getDuration(): void {}
  playVideo(): void {}
  getPlaylistIndex(): void {}
  getVideoData(): void {}
  getPlayerState(): void {}
  _triggerOnReady(): void {
    this.opts.events.onReady({ target: this });
  }
  _triggerOnStateChange(data: number) {
    this.opts.events.onStateChange({ data });
  }
  _triggerOnError(err: any) {
    this.opts.events.onError(err);
  }
}

MockPlayer.prototype.playVideoAt = jest.fn();
MockPlayer.prototype.getCurrentTime = jest.fn();
MockPlayer.prototype.getDuration = jest.fn();
MockPlayer.prototype.playVideo = jest.fn();
MockPlayer.prototype.getPlaylistIndex = jest.fn();
MockPlayer.prototype.getVideoData = jest.fn();
MockPlayer.prototype.getPlayerState = jest.fn();

global['YT'] = {
  Player: MockPlayer
};

import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { YouTubePlayer, YouTubePlayerProps } from '../YouTubePlayer';

describe('test Player component', () => {
  let props: YouTubePlayerProps;

  beforeEach(() => {
    props = {
      currentPlaylist: undefined,
      currentQueue: [],
      nextSong: jest.fn(),
      nextStatus: undefined,
      playerStatus: 'stopped',
      queueIndex: 0,
      setProgress: jest.fn(),
      setSongInfo: jest.fn(),
      updatePlayerStatus: jest.fn(),
      updateQueueIndex: jest.fn()
    };
  });

  test('handles default', () => {
    const component = renderer.create(<YouTubePlayer {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
