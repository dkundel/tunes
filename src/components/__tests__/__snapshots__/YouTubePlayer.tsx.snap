// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Player component handles default 1`] = `
<div
  id="youTubePlayer"
/>
`;
